### ISSUE TITLE
[F#] Implement new Concept Exercise: recursion-advanced

### ISSUE THREAD

**Issue by ErikSchierboom** [exercism/v3#956]
This issue describes how to implement the `recursion-advanced` concept exercise for the F# track.

## Goal

The goal of this exercise is to teach the student some advanced parts of the Concept of Recursion in [F#][recursion]. It will show how to work with non-trivial recursive types that have many cases and where cases' data refers to multiple values of the recursive type themselves. This will make it impossible to write trivially tail-recursive functions. The exercise should show that this _is_ possible using continuations.

It would also be great if we could find a way to also demonstrate mutually recursive functions in the exercise, but if that does not work we can mention it in the documentation

Examples of recursive types that could be used here are binary trees (where nodes have two child trees), a file system or modelling mathematical or boolean expressions. The latter could look like this:

```fsharp
type BooleanExpression =
    | Constant of bool
    | And of BooleanExpression * BooleanExpression
    | Or of BooleanExpression * BooleanExpression
    | Not of BooleanExpression
```

Note that the above are just suggestions; if you have a better idea, please use it!

## Learning objectives

- Know how to work with non-trivial recursive types.
- Know how to use continuations to write tail-recursive functions to handle non-trivial recursive types.
- Mutually recursive functions.

## Out of scope

Nothing.

## Concepts

The Concepts this exercise unlocks are:

- `recursion-advanced`: know how to work with non-trivial recursive types; know how to use continuations to write tail-recursive functions to handle non-trivial recursive types.

## Prerequisites

This exercise's prerequisites Concepts are:

- `recursion-basic`: know how to define recursive functions and types.
- `tuples-basic`: know how to use tuples as the recursive type will use them.

### Hints

- [Recursion in F#][recursion-and-tail-recursion]: exploration of recursion in F#, includes continuation passing.
- [Advanced recursion techniques in F#][advanced-recursion]: shows some advanced recursion techniques including continuation passing.
- [Mutually recursive functions][mutually-recursive-functions]: shows how to define mutually recursive functions.

### After

- [Recursion in F#][recursion-and-tail-recursion]: exploration of recursion in F#, includes continuation passing.
- [Advanced recursion techniques in F#][advanced-recursion]: shows some advanced recursion techniques including continuation passing.
- [Mutually recursive functions][mutually-recursive-functions]: shows how to define mutually recursive functions.

We could also mention the fact that C# does _not_ have tail recursion, which can be very interesting for any C# developers that want to use recursive functions in their C# code.

## Representer

This exercise does not require any specific representation logic to be added to the [representer][representer].

## Analyzer

This exercise does not require any specific logic to be added to the [analyzer][analyzer].

## Implementing

To implement the concept exercise, the following files must be created:

<pre>
languages
└── fsharp
    └── exercises
        └── concept
            └── recursion-advanced
                ├── .docs
                |   ├── instructions.md
                |   ├── introduction.md
                |   ├── hints.md
                |   └── after.md
                ├── .meta
                |   ├── config.json
                |   |── design.md
                |   └── Example.fs
                ├── RecursionAdvanced.fs
                ├── RecursionAdvanced.fsproj
                └── RecursionAdvancedTests.fs
</pre>

## Step 1: add .docs/introduction.md

This file contains an introduction to the concept. It should be explicit about what the student should learn from the exercise, and provide a short, concise introduction to the concept(s). The aim is to give the student just enough context to figure things out themselves and solve the exercise, as research has shown that self-discovery is the most effective learning experience. Mentioning technical terms that the student can Google if they so want, is preferable over including any code samples or an extensive description. For example we might describe a string as a "Sequence of Unicode characters" or a "series of bytes" or "an object". Unless the student needs to understand the details of what those mean to be able to solve the exercise we should not give more info in this introduction - instead allowing the student to Google, ignore, or map their existing knowledge.

## Step 2: add .docs/instructions.md

This file contains instructions for the exercise. It should explicitly explain what the student needs to do (define a method with the signature `X(...)` that takes an A and returns a Z), and provide at least one example usage of that function. If there are multiple tasks within the exercise, it should provide an example of each.

## Step 3: add .docs/hints.md

If the student gets stuck, we will allow them to click a button requesting a hint, which shows this file. This will not be a "recommended" path and we will (softly) discourage them using it unless they can't progress without it. As such, it's worth considering that the student reading it will be a little confused/overwhelmed and maybe frustrated.

The file should contain both general and task-specific "hints". These hints should be enough to unblock almost any student. They might link to the docs of the functions that need to be used.

The hints should not spell out the solution, but instead point to a resource describing the solution (e.g. linking to documentation for the function to use).

## Step 4: add .docs/after.md

Once the student completes the exercise they will be shown this file, which should provide them with a summary of what the exercise aimed to teach. This document can also link to any additional resources that might be interesting to the student in the context of the exercise.

These files are also all described in the [concept exercises document][docs-concept-exercises].

## Step 5: update languages/fsharp/config.json

An entry should be added to the track's `config.json` file for the new concept exercise:

```json
{
  ...
  "exercises": {
    "concept": [
      ...
      {
        "slug": "recursion-advanced",
        "uuid": "ac890472-69f3-11ea-bc55-0242ac130003",
        "concepts": ["recursion-advanced"],
        "prerequisites": ["recursion-basic"]
      }
    ]
  }
}
```

## Step 6: adding track-specific files

These files are specific to the F# track:

- `RecursionAdvanced.fs`. the stub implementation file, which is the starting point for students to work on the exercise.
- `RecursionAdvanced.fsproj`: the F# project file.
- `RecursionAdvancedTests.fs`: the test suite.
- `.meta/Example.fs`: an example implementation that passes all the tests. The project file should _not_ include this file.

## Step 7: update the general concept document

Add the exercise to the [concept's shared document's][reference] `## Implementations` section ([example](https://github.com/exercism/v3/blob/master/reference/types/string.md#implementations)).

## Step 8: updating list of implemented exercises

- Add the exercise to the [list of implemented exercises][implemented-exercises].

## Step 9: add .meta/design.md:

This file contains information on the exercise's design, which includes things like its goal, its teaching goals, what not to teach, and more ([example][meta-design]). This information can be extracted from this GitHub issue.

## Step 10: add .meta/config.json:

This file contains meta information on the exercise, which currently only includes the exercise's contributors ([example][meta-config.json]).

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue.

[how-to-implement-a-concept-exercise]: https://github.com/exercism/v3/blob/master/docs/maintainers/generic-how-to-implement-a-concept-exercise.md
[implemented-exercises]: https://github.com/exercism/v3/tree/master/languages/fsharp/exercises/concept/README.md#implemented-exercises
[reference]: https://github.com/exercism/v3/blob/master/languages/fsharp/reference/README.md#reference-docs
[analyzer]: https://github.com/exercism/fsharp-analyzer
[representer]: https://github.com/exercism/fsharp-representer
[meta-design]: https://github.com/exercism/v3/blob/master/languages/fsharp/exercises/concept/numbers/.meta/design.md
[meta-config.json]: https://github.com/exercism/v3/blob/master/languages/fsharp/exercises/concept/numbers/.meta/config.json
[docs-concept-exercises]: https://github.com/exercism/v3/blob/master/docs/concept-exercises.md
[recursion]: https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/functions/recursive-functions-the-rec-keyword
[recursion-and-tail-recursion]: https://cyanbyfuchsia.wordpress.com/2014/02/12/recursion-and-tail-recursion-in-f/
[advanced-recursion]: https://www.gresearch.co.uk/article/advanced-recursion-techniques-in-f/
[mutually-recursive-functions]: https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/functions/recursive-functions-the-rec-keyword#mutually-recursive-functions
